nightly cross build started failing 2025-08-06

### Summary

I have a GitHub Actions workflow that cross-builds a nightly rust toolchain from x86_64-pc-windows-gnu to x86_64-pc-cygwin.  On 2025-08-06, this started failing.

### Command used

```sh
  PKG_CONFIG_SYSROOT_DIR="$(cygpath -w /)" \
  PKG_CONFIG="$(cygpath -w /usr/bin/pkgconf.exe)" \
  LIBGIT2_NO_VENDOR=1 \
  OPENSSL_DIR="$(cygpath -w /usr)" \
  OPENSSL_LIB_DIR="$(cygpath -w /usr/lib)" \
  OPENSSL_STATIC=0 \
  DESTDIR="$(cygpath -w "$PWD/build-Cygwin/dest-rust")" \
  python x.py install --stage 2
```

### Expected behaviour

successful build, as on 2025-08-05 and before

### Actual behaviour

```
Building stage2 tool cargo (x86_64-pc-windows-gnu -> x86_64-pc-cygwin)
     Compiling proc-macro2 v1.0.95
     Compiling unicode-ident v1.0.18
     Compiling libc v0.2.174
     Compiling cfg-if v1.0.1
  error[E0463]: can't find crate for `core`
    |
    = note: the `x86_64-pc-windows-gnu` target may not be installed
    = help: consider downloading the target with `rustup target add x86_64-pc-windows-gnu`
    = help: consider building the standard library from source with `cargo build -Zbuild-std`
  
  error[E0463]: can't find crate for `std`
    |
    = note: the `x86_64-pc-windows-gnu` target may not be installed
    = help: consider downloading the target with `rustup target add x86_64-pc-windows-gnu`
    = help: consider building the standard library from source with `cargo build -Zbuild-std`
  
  For more information about this error, try `rustc --explain E0463`.
```

### Bootstrap configuration (bootstrap.toml)
```toml
# see src/bootstrap/defaults/
profile = "dist"

# see src/bootstrap/src/utils/change_tracker.rs
change-id = 138986

[build]
submodules = false
build = "x86_64-pc-windows-gnu"
host = ["x86_64-pc-cygwin"]
target = [
  "x86_64-pc-cygwin",
]
build-dir = "build-Cygwin"
python = "$MSYS2_ROOT/mingw64/bin/python.exe"
locked-deps = true
vendor = true
tools = [
"clippy", "rustdoc", "rustfmt", "rust-analyzer-proc-macro-srv", "analysis",
  "cargo",
]
profiler = true
docs = false

[install]
prefix = "/usr"
sysconfdir = "/etc"

[llvm]
link-shared = true

[rust]
codegen-units-std = 1
channel = "stable"
rpath = false
frame-pointers = true
llvm-bitcode-linker = false
lld = false
llvm-tools = false
codegen-tests = false
deny-warnings = false
backtrace-on-ice = true
control-flow-guard = true
lto = "thin"

[dist]
compression-formats = ["gz"]
compression-profile = "fast"
include-mingw-linker = false

[target.x86_64-pc-windows-gnu]
cc = "$MSYS2_ROOT/mingw64/bin/clang.exe"
cxx = "$MSYS2_ROOT/mingw64/bin/clang++.exe"
ar = "$MSYS2_ROOT/mingw64/bin/llvm-ar.exe"
ranlib = "$MSYS2_ROOT/mingw64/bin/llvm-ranlib.exe"
linker = "$MSYS2_ROOT/mingw64/bin/clang.exe"
llvm-config = "$MSYS2_ROOT/mingw64/bin/llvm-config.exe"
crt-static = true

[target.x86_64-pc-cygwin]
cc = "$WORKSPACE/gcc.exe"
cxx = "$WORKSPACE/g++.exe"
ar = "$WORKSPACE/ar.exe"
ranlib = "$WORKSPACE/ranlib.exe"
linker = "$WORKSPACE/x86_64-pc-cygwin-gcc.exe"
llvm-config = "$WORKSPACE/llvm-config.exe"
profiler = false
```

### Operating system

windows-latest

### HEAD

https://static.rust-lang.org/dist/2025-08-06/rustc-nightly-src.tar.xz

### Additional context

https://github.com/jeremyd2019/cygwin-rust-bootstrap/actions/runs/16774967434/job/47498825933 - all the scripts and configs are in that repository.

<!--
Include the complete build log in the section below.
Enable backtrace and verbose mode if possible for more detailed information e.g., with `RUST_BACKTRACE=1 ./x build -v`.
-->
<details><summary>Build Log</summary>
<p>

too long to paste, see https://github.com/jeremyd2019/cygwin-rust-bootstrap/actions/runs/16774967434/job/47498825933

</p>
</details>

--- a/src/bootstrap/src/core/build_steps/tool.rs
+++ b/src/bootstrap/src/core/build_steps/tool.rs
@@ -334,7 +334,8 @@ pub fn prepare_tool_cargo(
 /// Determines how to build a `ToolTarget`, i.e. which compiler should be used to compile it.
 /// The compiler stage is automatically bumped if we need to cross-compile a stage 1 tool.
 pub enum ToolTargetBuildMode {
-    /// Build the tool using rustc that corresponds to the selected CLI stage.
+    /// Build the tool for the given `target` using rustc that corresponds to the top CLI
+    /// stage.
     Build(TargetSelection),
     /// Build the tool so that it can be attached to the sysroot of the passed compiler.
     /// Since we always dist stage 2+, the compiler that builds the tool in this case has to be
@@ -366,7 +367,10 @@ pub(crate) fn get_tool_target_compiler(
     } else {
         // If we are cross-compiling a stage 1 tool, we cannot do that with a stage 0 compiler,
         // so we auto-bump the tool's stage to 2, which means we need a stage 1 compiler.
-        builder.compiler(build_compiler_stage.max(1), builder.host_target)
+        let build_compiler = builder.compiler(build_compiler_stage.max(1), builder.host_target);
+        // We also need the host stdlib to compile host code (proc macros/build scripts)
+        builder.std(build_compiler, builder.host_target);
+        build_compiler
     };
     builder.std(compiler, target);
     compiler

--- a/src/bootstrap/src/core/builder/tests.rs
+++ b/src/bootstrap/src/core/builder/tests.rs
@@ -976,6 +976,21 @@ mod snapshot {
     }
 
     #[test]
+    fn build_cargo_cross() {
+        let ctx = TestCtx::new();
+        insta::assert_snapshot!(
+            ctx.config("build")
+                .paths(&["cargo"])
+                .hosts(&[TEST_TRIPLE_1])
+                .render_steps(), @r"
+        [build] llvm <host>
+        [build] rustc 0 <host> -> rustc 1 <host>
+        [build] rustc 1 <host> -> std 1 <target1>
+        [build] rustc 1 <host> -> cargo 2 <target1>
+        ");
+    }
+
+    #[test]
     fn dist_default_stage() {
         let ctx = TestCtx::new();
         assert_eq!(ctx.config("dist").path("compiler").create_config().stage, 2);

--- a/src/bootstrap/src/core/builder/tests.rs
+++ b/src/bootstrap/src/core/builder/tests.rs
@@ -985,6 +985,7 @@ mod snapshot {
                 .render_steps(), @r"
         [build] llvm <host>
         [build] rustc 0 <host> -> rustc 1 <host>
+        [build] rustc 1 <host> -> std 1 <host>
         [build] rustc 1 <host> -> std 1 <target1>
         [build] rustc 1 <host> -> cargo 2 <target1>
         ");

